src/Compiler/NodeVisitor/NodeVisitorStorage.php

```diff
@@ -9,26 +9,34 @@
 
 final class NodeVisitorStorage
 {
-    /** @var array<int, NodeVisitor[]> */
-    private array $nodeVisitors = [];
+    private const WITHOUT_SCOPE = 'without_scope';
 
-    public function addNodeVisitor(int $priority, NodeVisitor $nodeVisitor): void
+    private const WITH_SCOPE = 'with_scope';
+
+    /** @var array<string, array<int, NodeVisitor[]>> */
+    private array $nodeVisitors = [
+        self::WITHOUT_SCOPE => [],
+        self::WITH_SCOPE => [],
+    ];
+
+    public function addNodeVisitor(int $priority, NodeVisitor $nodeVisitor, bool $withScope = false): void
```

Comment:
add check if there is interface when it should be

Reply:
Would not it be better to have two methods or differentiate based on visitor implementing ExprTypeNodeVisitorInterface?

Reply:
Yes I can use the interface as a flag